subsys/bluetooth/mesh/adv_ext.c

```diff
@@ -250,116 +246,151 @@ static int adv_send(struct bt_mesh_ext_adv *ext_adv, struct bt_mesh_adv *adv)
-
 	while ((adv = bt_mesh_adv_get_by_tag(ext_adv->tags, K_NO_WAIT))) {
 		/* busy == 0 means this was canceled */
 		if (!adv->ctx.busy) {
 			bt_mesh_adv_unref(adv);
 			continue;
 		}
 
+		if (!stop_proxy_adv(ext_adv)) {
+			LOG_WRN("Advertising %p canceled due to proxy adv failed to stop", adv);
+			bt_mesh_adv_send_start(0, -ECANCELED, &adv->ctx);
+			bt_mesh_adv_unref(adv);
+			continue;
+		}
+
 		adv->ctx.busy = 0U;
 		err = adv_send(ext_adv, adv);
 
 		bt_mesh_adv_unref(adv);
 
 		if (!err) {
-			return; /* Wait for advertising to finish */
+			return true; /* Wait for advertising to finish */
 		}
 	}
 
+	return false;
+}
+
+static void start_proxy_sol_or_proxy_adv(struct bt_mesh_ext_adv *ext_adv)
+{
 	if (ext_adv->instance == NULL) {
 		LOG_DBG("Advertiser is suspended or deleted");
 		return;
 	}
 
-	if (IS_ENABLED(CONFIG_BT_MESH_PROXY_SOLICITATION) &&
-	    !bt_mesh_sol_send()) {
+	if (!(ext_adv->tags & BT_MESH_ADV_TAG_BIT_PROXY)) {
 		return;
 	}
 
-	if (!IS_ENABLED(CONFIG_BT_MESH_GATT_SERVER) ||
-	    !(ext_adv->tags & BT_MESH_ADV_TAG_BIT_PROXY)) {
+	if (IS_ENABLED(CONFIG_BT_MESH_PROXY_SOLICITATION)) {
+		if (!stop_proxy_adv(ext_adv)) {
+			return;
+		}
+
+		if (!bt_mesh_sol_send()) {
+			return;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_MESH_GATT_SERVER) &&
+	    !atomic_test_and_set_bit(ext_adv->flags, ADV_FLAG_PROXY)) {
+		if (bt_mesh_adv_gatt_send()) {
+			atomic_clear_bit(ext_adv->flags, ADV_FLAG_PROXY);
+			return;
+		}
+	}
+}
+
+static void send_pending_adv(struct k_work *work)
+{
+	static const char * const adv_tag_to_str[] = {
+		[BT_MESH_ADV_TAG_LOCAL]  = "local",
+		[BT_MESH_ADV_TAG_RELAY]  = "relay",
+		[BT_MESH_ADV_TAG_PROXY]  = "proxy",
+		[BT_MESH_ADV_TAG_FRIEND] = "friend",
+		[BT_MESH_ADV_TAG_PROV]   = "prov",
+	};
+	struct bt_mesh_ext_adv *ext_adv;
+	bool sent;
+
+	ext_adv = CONTAINER_OF(work, struct bt_mesh_ext_adv, work);
+
+	if (atomic_test_bit(ext_adv->flags, ADV_FLAG_SUSPENDING)) {
+		LOG_DBG("Advertiser is suspending");
 		return;
 	}
 
-	atomic_set_bit(ext_adv->flags, ADV_FLAG_PROXY_START);
+	if (atomic_test_and_clear_bit(ext_adv->flags, ADV_FLAG_SENT)) {
+		LOG_DBG("Advertising stopped after %u ms for %s adv",
+			k_uptime_get_32() - ext_adv->timestamp,
+			ext_adv->adv ? adv_tag_to_str[ext_adv->adv->ctx.tag]
+				     : adv_tag_to_str[BT_MESH_ADV_TAG_PROXY]);
+
+		atomic_clear_bit(ext_adv->flags, ADV_FLAG_ACTIVE);
+		atomic_clear_bit(ext_adv->flags, ADV_FLAG_PROXY);
 
-	if (!bt_mesh_adv_gatt_send()) {
-		atomic_set_bit(ext_adv->flags, ADV_FLAG_PROXY);
+		if (ext_adv->adv) {
+			struct bt_mesh_adv_ctx ctx = ext_adv->adv->ctx;
+
+			ext_adv->adv->ctx.started = 0;
+			bt_mesh_adv_unref(ext_adv->adv);
+			bt_mesh_adv_send_end(0, &ctx);
+
+			ext_adv->adv = NULL;
+		}
 	}
 
-	if (atomic_test_and_clear_bit(ext_adv->flags, ADV_FLAG_SCHEDULE_PENDING)) {
-		schedule_send_with_mask(ext_adv, K_WORK_RUNNING);
+	sent = adv_queue_send_process(ext_adv);
+	if (sent) {
+		return;
 	}
+
+	start_proxy_sol_or_proxy_adv(ext_adv);
```

Comment:
It is a bit difficult to read the entire logic in `send_pending_adv` because some functions return `true` on success, some `false`. I suggest to align semantics of the return values of these functions. The easiest way is to return int in all of them (`stop_proxy_adv`, `adv_queue_send_process`, `start_proxy_sol_or_proxy_adv`).